fix: don't mkdir directories that already exist

TheCSDev opened the pull request:
This pull request aims to resolve #6412, by making checks for whether a project's working directory already exists, before attempting to `mkdir` it.  
This effectively prevents `reify.js` from accidentally executing `mkdir` on root drive letters on Windows, like `mkdir "D:/"` for example, as the Windows operating system prohibits such operations.  

The resolution of #6412 will inevitably introduce the `npm error Tracker "idealTree" already exists` issue, however I believe that is an issue separate from #6412, and that separate issue was already documented before, #7596 being a good example of one of those documentations (in my opinion).

As can be seen in #6412, this pull request makes changes to [the following line of code](https://github.com/npm/cli/blob/780afc50e3a345feb1871a28e33fa48235bc3bd5/workspaces/arborist/lib/arborist/reify.js#L126) in `reify.js`.

Of course, should one attempt to execute `npm i` in a drive letter that does not exist, then this error in my opinion **should** take place, informing the user that said drive letter and the path within cannot be created, and that error **will** take place. However this error in my opinion **should not** take place on existing drives, hence this pull request.

### Additional info
- **Affected file:** `/workspaces/arborist/lib/arborist/reify.js` line `126`
- **Relevant issue:** #6412
- **Other related issues/prs:** #7596

TheCSDev committed:
Update reify.js

Resolution attempt for #6412.

Unfortunately, resolution of that issue introduces another error; `Tracker "idealTree" already exists`.

wraithgar commented on workspaces/arborist/lib/arborist/reify.js:
```diff
@@ -12,6 +12,7 @@ const hgi = require('hosted-git-info')
 const rpj = require('read-package-json-fast')
 
 const { dirname, resolve, relative, join } = require('node:path')
+const { existsSync } = require("node:fs")
```
We should use `exists` from the already imported `node:fs/promises` and await it.

TheCSDev replied:
I would like to make this correction, however I am failing to locate a definition for `exists` in the `node:fs/promises` module.  
At first VScode intellisense was not showing any definitions for the given method (although it did find `exists` in `node:fs`), so I decided to double-check in NodeJS runtime. Below is the output;
```bat
C:\Users\User>node --version
v20.18.0

C:\Users\User>node
Welcome to Node.js v20.18.0.
Type ".help" for more information.
> require("node:fs/promises").exists
undefined
>
```

wraithgar replied:
Ah that's right, `exists` is deprecated.  `fs.stat` or `fs.access` is recommended instead, depending on the context.  `access` would be best here as we do not need the full stat info.

TheCSDev replied:
I have made another commit, this time replacing `existsSync` with `access` from `node:fs/promises`.
In this commit, if `access` raises an "ENOENT" `Error`, the directory is created. Any other types of access errors that are raised, are allowed to bubble up.

TheCSDev committed:
Update reify.js

Addressed style and syntax issues as per GitHub Workflows feedback.

TheCSDev committed:
Update reify.js

Now using `access` from `node:fs/promises` instead of `existsSync` from `node:fs`.

wraithgar commented:
`npm run lintfix` should fix linting errors.

TheCSDev committed:
Update reify.js

Now using single-quotes instead of double-quotes.

TheCSDev commented:
Ah, it appears I accidentally used double-quotes instead of single-quotes for the string.  
Also for some reason `npm run lintfix` did not spot any errors or make any "fixing" changes for me, so it took a bit of figuring out. I eventually found the `npm run lint -w @npmcli/arborist` command from a GitHub Workflows task, and that one was able to spot the syntax error I made.  
I will now make one more commit to correct that error, and hopefully that should be it. Thank you for helping me.

wraithgar commented:
Oh shoot, sorry I forgot this was the cli repo. Yeah you gotta either be in the arborist directory or specify the workspace.  I'll have to remember that when suggesting that in the future.

wraithgar commented on workspaces/arborist/lib/arborist/reify.js:
```diff
@@ -123,10 +124,19 @@ module.exports = cls => class Reifier extends cls {
+        await access(resolvedPath)
+      } catch (accessError) {
+        if (accessError.code === 'ENOENT') {
+          await mkdir(resolvedPath, { recursive: true })
```
Just a note for future devs looking at this change.  Any error other than `ENOENT` that `access` gets is going to also be thrown if attempting `mkdir`.  I think only ignoring that error is the correct one.

wraithgar commented on workspaces/arborist/lib/arborist/reify.js:
```diff
@@ -123,10 +124,19 @@ module.exports = cls => class Reifier extends cls {
+        if (accessError.code === 'ENOENT') {
+          await mkdir(resolvedPath, { recursive: true })
+        } else {
+          throw accessError
```
The reason tests are failing is because there aren't any that test for this scenario, i.e. an error other than `ENOENT` happening.

TheCSDev replied:
Will any changes have to be made to this code to resolve the failing tests in this case?

wraithgar replied:
A test will have to be added to cover that code path.

owlstronaut commented:
Closing due to inactivity / merge conflicts / incomplete work+tests — feel free to reopen or submit a fresh PR after addressing these. Thanks for contributing!
